effects/push-cache/default.nix

```diff
@@ -0,0 +1,160 @@
+{ lib, withSystem, config, ... }:
+let pkgs-x86_64-linux = withSystem "x86_64-linux" ({ pkgs, ... }: pkgs);
+in {
+  imports = [ ../../flake-module.nix ];
+
+  options = {
+    push-cache-effect = {
+      enable = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Enables an effect that pushes certain outputs to a different binary cache.
+
+          Hercules CI normally pushes everything to the cache(s) configured on the agent. This effect supplements that behavior by letting you push a subset of those to a different cache.
+          Note that it only pushes the output closure, and not the closures of build dependencies used during the build stage of the CI job. (Unless those closures happen to also be part of the output or "runtime" closure)
+        '';
+      };
+      attic-client-pkg = lib.mkOption {
+        type = lib.types.package;
+        description = ''
+          Version of the attic-client package to use on \"x86_64-linux\".
+
+          Hint: You can use `attic.packages.x86_64-linux.attic-client` from the attic flake.
+        '';
+        default = pkgs-x86_64-linux.attic-client or (throw
+          "push-cache-effect.attic-client-pkg: It seems that attic hasn't been packaged in Nixpkgs (yet?). Please check <nixpkgs packaging request issue> or set <option> manually.");
+      };
+      cachix-pkg = lib.mkOption {
+        type = lib.types.package;
+        default = pkgs-x86_64-linux.cachix;
+        description =
+          ''Version of the cachix package to use on "x86_64-linux".'';
+      };
+      caches = lib.mkOption {
+        description = ''
+          An attribute set, each `name: value` pair translates to
+          an effect under `onPush.default.outputs.effects.push-cache-effect.name`.
+        '';
+        example = ''
+          {
+            our-cachix = {
+              type = \"cachix\";
+              secretName = \"our-cachix-token\";
+              branches = [ \"master\" ];
+              packages = [ pkgs.hello ];
+            };
+          }
+        '';
+        type = lib.types.attrsOf (lib.types.submodule ({ name, ... }: {
+          options = {
+            name = lib.mkOption {
+              type = lib.types.str;
+              default = name;
+              description = ''
+                Name of the effect. By default it's the attribute name.
+              '';
+            };
+            type = lib.mkOption {
+              type = lib.types.enum [ "attic" "cachix" ];
+              description = ''A string "attic" or "cachix".'';
+            };
+            packages = lib.mkOption {
+              type = with lib.types; listOf package;
+              description = "List of packages to push to the cache.";
+              example = "[ pkgs.hello ]";
```

Comment:
We could use` flakeToOutputs` from  `hercules-ci-agent` to match the behavior of the default build
https://github.com/hercules-ci/hercules-ci-agent/blob/0e9e67bb7d41ea2bbb13982887ff3b9c511a2aba/hercules-ci-agent/data/default-herculesCI-for-flake.nix#L48
or better combine `herculesCI.onPush.*.outputs` to match the hercules ci build phase 
https://docs.hercules-ci.com/hercules-ci-agent/evaluation#attribute-herculesCI.onPush-outputs

Reply:
Interesting! Didn't know `flakeToOutputs`

Reply:
This would involve writing a attribute tree traversal to match the agent behavior, with the right `recuresIntoAttrs` behavior, rejecting things with `_type`, etc.
That could be done, but I wouldn't encode into this module the assumption that every output should be pushed.

Note also that this effect can not replicate the whole agent behavior, such as pushing all the right build dependencies.
This means that it's not a replacement for the cache that's used by the agents transparently.
More significantly, it also means that this effect is not 100% suitable as a build cache or "development cache", so I doubt that e.g. the `checks` or the `devShell` dependencies should be pushed by it. A better use case for this effect is as a "release cache" - for only the end products, such as packages, or perhaps also paths for deployment, such as `nixosConfigurations` or the inputs of effects (if flake consumers use the same effect functions).

Reply:
Resolved: we leave it unspecified and the responsibility of the caller since replicating default builder behavior is infeasible.